src/Compilers/Core/Portable/MetadataReader/MetadataTypeName.cs

```diff
@@ -204,14 +266,27 @@ public string UnmangledTypeName
             {
                 if (_unmangledTypeName == null)
                 {
-                    Debug.Assert(_inferredArity == -1);
                     _unmangledTypeName = MetadataHelpers.InferTypeArityAndUnmangleMetadataName(TypeName, out _inferredArity);
                 }
 
                 return _unmangledTypeName;
             }
         }
 
+        public ReadOnlyMemory<char> UnmangledTypeNameMemory
+        {
+            get
+            {
+                if (_unmangledTypeNameMemory.Equals(default))
+                {
+                    Debug.Assert(_inferredArity == -1);
+                    _unmangledTypeNameMemory = MetadataHelpers.InferTypeArityAndUnmangleMetadataName(TypeNameMemory, out _inferredArity);
```

Comment:
>_unmangledTypeNameMemory = MetadataHelpers.InferTypeArityAndUnmangleMetadataName(TypeNameMemory, out _inferredArity); [](http://example.com/codeflow?start=20&length=117)

Could make a single call to InferTypeArityAndUnmangleMetadataName also

Reply:
> That would very much defeat the purpose of this PR. Because it would end up allocating the strings for _typeName and _namespaceName (which is waht we're trying to avoid here) :)

Duh Todd, that makes a lot of sense. Perhaps a better thought then would be to have the string versions defer to the memory versions inside the null check and ToString their resultant values to _typeName and _namespaceName? Or is an instance of this class not going to be typically queried for both the ReadOnlyMemory<char> and string data?